### PR TITLE
51423: wp_privacy_generate_personal_data_export_file fixes type match for array_merge (PHP 8)

### PR DESCRIPTION
**Pull request by hellofromtonya**
Trac ticket: https://core.trac.wordpress.org/ticket/51423

Function: `wp_privacy_generate_personal_data_export_file()`

###  Problems
When/if the `'_export_data_grouped'` post meta exists but is not an array, a PHP Warning (< 8)/Error (8+) is thrown for `array_merge` argument 2 type mismatch.

See in action https://3v4l.org/9gLZ8

#### PHP Warning/Error

- PHP 8+: 
```
PHP Fatal error:  Uncaught TypeError: array_merge(): Argument #2 must be of type array, string given in .../public/wp-admin/includes/privacy-tools.php:397
```

- PHP 5.6 to 5.7.x:
```
PHP Warning:  array_merge(): Expected parameter 2 to be an array, string given in .../public/wp-admin/includes/privacy-tools.php on line 397
```

- and PHP 7.2 to 7.4.x:
```
PHP Warning:  count(): Parameter must be an array or an object that implements Countable in .../public/wp-admin/includes/privacy-tools.php on line 399
```

### Current Behavior
Case: `'_export_data_grouped'` post meta is not an array

```php
$groups = array_merge( array( 'about' => $about_group ), $groups );
// $groups = null

$groups_count = count( $groups );
// $groups_count = 0

$groups_json = wp_json_encode( $groups );
// $groups_json = 'null'
```

With these values, the following happens:

- The export file is created
- But the groups TOC or groups content

The resultant zip file looks like this:

- `export.json` file:
```json
{"Personal Data Export for tester2@test.com":null}
```

- `index.html` file:
```html
<!DOCTYPE html>
<html>
<head>
<meta http-equiv='Content-Type' content='text/html; charset=UTF-8' />
<style type='text/css'>body { color: black; font-family: Arial, sans-serif; font-size: 11pt; margin: 15px auto; width: 860px; }table { background: #f0f0f0; border: 1px solid #ddd; margin-bottom: 20px; width: 100%; }th { padding: 5px; text-align: left; width: 20%; }td { padding: 5px; }tr:nth-child(odd) { background-color: #fafafa; }.return-to-top { text-align: right; }</style><title>Personal Data Export for tester2@test.com</title></head>
<body>
<h1 id="top">Personal Data Export</h1></body>
</html>
```

### Fix

Check if the post meta is an array. 

- If yes, continue with the `array_merge` and `count`
- Else:
     - invoke `_doing_it_wrong` (to alert the developer where the problem exists while eliminating the warning/error)
     - set the variables to retain the current behavior and file generations
 
See it in action https://3v4l.org/P3VAM.

See 
---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: src/wp-admin/includes/privacy-tools.php [hunk 1]
```diff
@@ -381,19 +378,35 @@ function wp_privacy_generate_personal_data_export_file( $request_id ) {
...
+	} elseif ( '' === $groups || false === $groups ) {
+		$groups       = array( 'about' => $about_group );
+		$groups_count = 1;
+	} else {
+		wp_send_json_error(
+			sprintf(
+				'Warning: array_merge(): Expected parameter 2 to be an array, %1$s given.',
+				gettype( $groups )
+			)
+		);
```
Comment (hellofromtonya) [inline src/wp-admin/includes/privacy-tools.php:397]: Preserves the original warning when/if `'_export_data_grouped'` exists and is not an `array` type.

- < PHP 7.3
>Warning: array_merge(): Argument #2 is not an array

- PHP 7.3-7.4
>Warning: array_merge(): Expected parameter 2 to be an array, string given

- PHP 8
>Fatal error: Uncaught TypeError: array_merge(): Argument #2 must be of type array, string given

See it in action here https://3v4l.org/VlCAl

File: src/wp-admin/includes/privacy-tools.php [hunk 2]
```diff
@@ -388,14 +385,32 @@ function wp_privacy_generate_personal_data_export_file( $request_id ) {
...
```
Comment (hellofromtonya) [inline src/wp-admin/includes/privacy-tools.php:404]: Outline of the code to change, though the doing it wrong needs attention.

```suggestion
	// False will be returned when the request ID is invalid.
	if ( false === $groups ) {
		$groups       = array( 'about' => $about_group );
		$groups_count = 1;
	} else {
		if ( ! is_array ( $groups ) ) {
			// @TODO: Update this message.
			__doing_it_wrong(
				'',
				sprintf(
					'Warning: array_merge(): Expected parameter 2 to be an array, %1$s given.',
					gettype( $groups )
				)
			);
		}
		$groups            = array_merge( array( 'about' => $about_group ), (array) $groups );
		$groups_count = count( $groups );
	}			
```

**Comment by hellofromtonya** [comment]
## Testing BEFORE fixes

### PHP 7.3.5

Results: 
- Export file downloaded ✅ 
- 2 PHP Warnings

Screen Shot 2021-03-18 at 4 46 23 PM

Error log:
```
[18-Mar-2021 21:46:02 UTC] PHP Warning:  array_merge(): Expected parameter 2 to be an array, string given in .../public/wp-admin/includes/privacy-tools.php on line 397
[18-Mar-2021 21:46:02 UTC] PHP Stack trace:
[18-Mar-2021 21:46:02 UTC] PHP   1. {main}() .../public/wp-admin/admin-ajax.php:0
[18-Mar-2021 21:46:02 UTC] PHP   2. do_action() .../public/wp-admin/admin-ajax.php:187
[18-Mar-2021 21:46:02 UTC] PHP   3. WP_Hook->do_action() .../public/wp-includes/plugin.php:484
[18-Mar-2021 21:46:02 UTC] PHP   4. WP_Hook->apply_filters() .../public/wp-includes/class-wp-hook.php:316
[18-Mar-2021 21:46:02 UTC] PHP   5. wp_ajax_wp_privacy_export_personal_data() .../public/wp-includes/class-wp-hook.php:292
[18-Mar-2021 21:46:02 UTC] PHP   6. apply_filters() .../public/wp-admin/includes/ajax-actions.php:4911
[18-Mar-2021 21:46:02 UTC] PHP   7. WP_Hook->apply_filters() .../public/wp-includes/plugin.php:212
[18-Mar-2021 21:46:02 UTC] PHP   8. wp_privacy_process_personal_data_export_page() .../public/wp-includes/class-wp-hook.php:292
[18-Mar-2021 21:46:02 UTC] PHP   9. do_action() .../public/wp-admin/includes/privacy-tools.php:844
[18-Mar-2021 21:46:02 UTC] PHP  10. WP_Hook->do_action() .../public/wp-includes/plugin.php:484
[18-Mar-2021 21:46:02 UTC] PHP  11. WP_Hook->apply_filters() .../public/wp-includes/class-wp-hook.php:316
[18-Mar-2021 21:46:02 UTC] PHP  12. wp_privacy_generate_personal_data_export_file() .../public/wp-includes/class-wp-hook.php:292
[18-Mar-2021 21:46:02 UTC] PHP  13. array_merge() .../public/wp-admin/includes/privacy-tools.php:397
[18-Mar-2021 21:46:02 UTC] PHP Warning:  count(): Parameter must be an array or an object that implements Countable in .../public/wp-admin/includes/privacy-tools.php on line 399
[18-Mar-2021 21:46:02 UTC] PHP Stack trace:
[18-Mar-2021 21:46:02 UTC] PHP   1. {main}() .../public/wp-admin/admin-ajax.php:0
[18-Mar-2021 21:46:02 UTC] PHP   2. do_action() .../public/wp-admin/admin-ajax.php:187
[18-Mar-2021 21:46:02 UTC] PHP   3. WP_Hook->do_action() .../public/wp-includes/plugin.php:484
[18-Mar-2021 21:46:02 UTC] PHP   4. WP_Hook->apply_filters() .../public/wp-includes/class-wp-hook.php:316
[18-Mar-2021 21:46:02 UTC] PHP   5. wp_ajax_wp_privacy_export_personal_data() .../public/wp-includes/class-wp-hook.php:292
[18-Mar-2021 21:46:02 UTC] PHP   6. apply_filters() .../public/wp-admin/includes/ajax-actions.php:4911
[18-Mar-2021 21:46:02 UTC] PHP   7. WP_Hook->apply_filters() .../public/wp-includes/plugin.php:212
[18-Mar-2021 21:46:02 UTC] PHP   8. wp_privacy_process_personal_data_export_page() .../public/wp-includes/class-wp-hook.php:292
[18-Mar-2021 21:46:02 UTC] PHP   9. do_action() .../public/wp-admin/includes/privacy-tools.php:844
[18-Mar-2021 21:46:02 UTC] PHP  10. WP_Hook->do_action() .../public/wp-includes/plugin.php:484
[18-Mar-2021 21:46:02 UTC] PHP  11. WP_Hook->apply_filters() .../public/wp-includes/class-wp-hook.php:316
[18-Mar-2021 21:46:02 UTC] PHP  12. wp_privacy_generate_personal_data_export_file() .../public/wp-includes/class-wp-hook.php:292
```
### PHP 8.0

Results:

- Fatal error
- Export failed

Screen Shot 2021-03-18 at 4 44 47 PM

Error log:
```
[18-Mar-2021 21:37:33 UTC] PHP Fatal error:  Uncaught TypeError: array_merge(): Argument #2 must be of type array, string given in .../public/wp-admin/includes/privacy-tools.php:397
Stack trace:
#0 .../public/wp-admin/includes/privacy-tools.php(397): array_merge(Array, 'invalid type')
#1 .../app/public/wp-includes/class-wp-hook.php(292): wp_privacy_generate_personal_data_export_file(21)
#2 .../public/wp-includes/class-wp-hook.php(316): WP_Hook->apply_filters(NULL, Array)
#3 .../public/wp-includes/plugin.php(484): WP_Hook->do_action(Array)
#4 .../public/wp-admin/includes/privacy-tools.php(844): do_action('wp_privacy_pers...', 21)
#5 .../public/wp-includes/class-wp-hook.php(292): wp_privacy_process_personal_data_export_page(Array, 3, 'tester2@test.co...', 1, 21, false, 'wordpress-media')
#6 .../public/wp-includes/plugin.php(212): WP_Hook->apply_filters(Array, Array)
#7 .../public/wp-admin/includes/ajax-actions.php(4911): apply_filters('wp_privacy_pers...', Array, 3, 'tester2@test.co...', 1, 21, false, 'wordpress-media')
#8 .../public/wp-includes/class-wp-hook.php(292): wp_ajax_wp_privacy_export_personal_data('')
#9 .../public/wp-includes/class-wp-hook.php(316): WP_Hook->apply_filters('', Array)
#10 .../public/wp-includes/plugin.php(484): WP_Hook->do_action(Array)
#11 .../public/wp-admin/admin-ajax.php(187): do_action('wp_ajax_wp-priv...')
#12 {main}
  thrown in .../public/wp-admin/includes/privacy-tools.php on line 397
```

### How Tested

Used this code in a must-use to change the meta value to a string  value:

```php
<?php

add_action( 'wp_privacy_personal_data_export_file', function( $request_id ) {
	update_post_meta( $request_id, '_export_data_grouped', 'invalid type' );
} , 5 );
```

**Tests to reproduce:**
1. Create a `wp-content/mu-plugins/51423.php` and add code above.
2. Add a test user, e.g. username `tester2` and password `tester2@test.com`.
3. Go to `Tools` > `Export Personal Data`.
4. Create a request for the user.
5. Attempt to download the export file.

File: src/wp-admin/includes/privacy-tools.php [hunk 3]
```diff
@@ -388,14 +385,32 @@ function wp_privacy_generate_personal_data_export_file( $request_id ) {
...
+				__FUNCTION__,
+				/* translators: %s: post meta key. */
+				sprintf( __( 'The %s post meta must be an array.' ), "<code>'_export_data_grouped'</code>" ),
+				'5.8.0'
+			);
+			$groups = (array) $groups;
+		}
+		$groups       = array_merge( array( 'about' => $about_group ), $groups );
+		$groups_count = count( $groups );
+	}
```
Comment (hellofromtonya) [inline src/wp-admin/includes/privacy-tools.php:405]: When post meta exists and is not an array, type casting and then merging causes a failure downstream. Why? The groups meta will not be in the right format as keys are not set. In this case, the problem is caused from here. The merging of non array means the post meta is ignored in < PHP 8. Therefore, skipping the type casting and merge when not an array is valid here.

https://gist.github.com/jrfnl/5f16fcc747c96339fa7133e0ac086e2d

Comment (hellofromtonya) [reply src/wp-admin/includes/privacy-tools.php]: Resolved by commit https://github.com/WordPress/wordpress-develop/pull/758/commits/ae220090decbd37d30493d0500520b47214eb80c

**Comment by hellofromtonya** [comment]
## Testing AFTER fixes

### PHP 7.3.5

Results: 
- Export file downloaded ✅ 
- No PHP Warnings
- 1 PHP Notice

Screen Shot 2021-03-22 at 2 18 05 PM


Error log:
```
[22-Mar-2021 19:16:05 UTC] PHP Notice:  wp_privacy_generate_personal_data_export_file was called <strong>incorrectly</strong>. The <code>'_export_data_grouped'</code> post meta must be an array. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 5.8.0.) in .../public/wp-includes/functions.php on line 5313
[22-Mar-2021 19:16:05 UTC] PHP Stack trace:
[22-Mar-2021 19:16:05 UTC] PHP   1. {main}() .../public/wp-admin/admin-ajax.php:0
[22-Mar-2021 19:16:05 UTC] PHP   2. do_action() .../public/wp-admin/admin-ajax.php:187
[22-Mar-2021 19:16:05 UTC] PHP   3. WP_Hook->do_action() .../public/wp-includes/plugin.php:484
[22-Mar-2021 19:16:05 UTC] PHP   4. WP_Hook->apply_filters() .../public/wp-includes/class-wp-hook.php:316
[22-Mar-2021 19:16:05 UTC] PHP   5. wp_ajax_wp_privacy_export_personal_data() /Users/hellofromtonya/Local/wpcore/app/public/wp-includes/class-wp-hook.php:292
[22-Mar-2021 19:16:05 UTC] PHP   6. apply_filters() .../public/wp-admin/includes/ajax-actions.php:4911
[22-Mar-2021 19:16:05 UTC] PHP   7. WP_Hook->apply_filters() .../public/wp-includes/plugin.php:212
[22-Mar-2021 19:16:05 UTC] PHP   8. wp_privacy_process_personal_data_export_page() .../public/wp-includes/class-wp-hook.php:292
[22-Mar-2021 19:16:05 UTC] PHP   9. do_action() .../public/wp-admin/includes/privacy-tools.php:862
[22-Mar-2021 19:16:05 UTC] PHP  10. WP_Hook->do_action() .../public/wp-includes/plugin.php:484
[22-Mar-2021 19:16:05 UTC] PHP  11. WP_Hook->apply_filters() .../public/wp-includes/class-wp-hook.php:316
[22-Mar-2021 19:16:05 UTC] PHP  12. wp_privacy_generate_personal_data_export_file() .../public/wp-includes/class-wp-hook.php:292
[22-Mar-2021 19:16:05 UTC] PHP  13. _doing_it_wrong() .../public/wp-admin/includes/privacy-tools.php:411
[22-Mar-2021 19:16:05 UTC] PHP  14. trigger_error() .../public/wp-includes/functions.php:5313
```

In the downloaded zip file:

- `export.json`:
```json
{"Personal Data Export for tester2@test.com":null}
```

-  `index.html
```html
<!DOCTYPE html>
<html>
<head>
<meta http-equiv='Content-Type' content='text/html; charset=UTF-8' />
<style type='text/css'>body { color: black; font-family: Arial, sans-serif; font-size: 11pt; margin: 15px auto; width: 860px; }table { background: #f0f0f0; border: 1px solid #ddd; margin-bottom: 20px; width: 100%; }th { padding: 5px; text-align: left; width: 20%; }td { padding: 5px; }tr:nth-child(odd) { background-color: #fafafa; }.return-to-top { text-align: right; }</style><title>Personal Data Export for tester2@test.com</title></head>
<body>
<h1 id="top">Personal Data Export</h1></body>
</html>
```

### PHP 8.0

Results:
- Export file downloaded ✅ 
- No PHP Fatal error
- 1 PHP Notice

Screen Shot 2021-03-22 at 2 23 30 PM


Error log:
```
[22-Mar-2021 19:23:22 UTC] PHP Notice:  wp_privacy_generate_personal_data_export_file was called <strong>incorrectly</strong>. The <code>'_export_data_grouped'</code> post meta must be an array. Please see <a href="https://wordpress.org/support/article/debugging-in-wordpress/">Debugging in WordPress</a> for more information. (This message was added in version 5.8.0.) in .../public/wp-includes/functions.php on line 5313
[22-Mar-2021 19:23:22 UTC] PHP Stack trace:
[22-Mar-2021 19:23:22 UTC] PHP   1. {main}().../public/wp-admin/admin-ajax.php:0
[22-Mar-2021 19:23:22 UTC] PHP   2. do_action() .../public/wp-admin/admin-ajax.php:187
[22-Mar-2021 19:23:22 UTC] PHP   3. WP_Hook->do_action() .../public/wp-includes/plugin.php:484
[22-Mar-2021 19:23:22 UTC] PHP   4. WP_Hook->apply_filters() .../public/wp-includes/class-wp-hook.php:316
[22-Mar-2021 19:23:22 UTC] PHP   5. wp_ajax_wp_privacy_export_personal_data() .../public/wp-includes/class-wp-hook.php:292
[22-Mar-2021 19:23:22 UTC] PHP   6. apply_filters() .../public/wp-admin/includes/ajax-actions.php:4911
[22-Mar-2021 19:23:22 UTC] PHP   7. WP_Hook->apply_filters() .../public/wp-includes/plugin.php:212
[22-Mar-2021 19:23:22 UTC] PHP   8. wp_privacy_process_personal_data_export_page() .../public/wp-includes/class-wp-hook.php:292
[22-Mar-2021 19:23:22 UTC] PHP   9. do_action() .../public/wp-admin/includes/privacy-tools.php:862
[22-Mar-2021 19:23:22 UTC] PHP  10. WP_Hook->do_action() .../public/wp-includes/plugin.php:484
[22-Mar-2021 19:23:22 UTC] PHP  11. WP_Hook->apply_filters() .../public/wp-includes/class-wp-hook.php:316
[22-Mar-2021 19:23:22 UTC] PHP  12. wp_privacy_generate_personal_data_export_file() .../public/wp-includes/class-wp-hook.php:292
[22-Mar-2021 19:23:22 UTC] PHP  13. _doing_it_wrong() .../public/wp-admin/includes/privacy-tools.php:411
[22-Mar-2021 19:23:22 UTC] PHP  14. trigger_error() .../public/wp-includes/functions.php:5313
```

In the downloaded zip file:

- `export.json`:
```json
{"Personal Data Export for tester2@test.com":null}
```

-  `index.html
```html
<!DOCTYPE html>
<html>
<head>
<meta http-equiv='Content-Type' content='text/html; charset=UTF-8' />
<style type='text/css'>body { color: black; font-family: Arial, sans-serif; font-size: 11pt; margin: 15px auto; width: 860px; }table { background: #f0f0f0; border: 1px solid #ddd; margin-bottom: 20px; width: 100%; }th { padding: 5px; text-align: left; width: 20%; }td { padding: 5px; }tr:nth-child(odd) { background-color: #fafafa; }.return-to-top { text-align: right; }</style><title>Personal Data Export for tester2@test.com</title></head>
<body>
<h1 id="top">Personal Data Export</h1></body>
</html>
```

**Comment by hellofromtonya** [comment]
@jrfnl This PR is ready for a final review. Since we last did a pair programming session on it, I discovered our solution caused a breaking change. Why? When the post meta does not exist or it's value is not an array, no about group content is generated in the downloaded zip file.

Why? See the updated notes in the description of this PR which includes 3v4l links to show it in action.

What's different since you and I last worked on it?

- See here https://github.com/WordPress/wordpress-develop/pull/758/commits/24839151602aff935b89f2f0ac14cf9f33a58985
- Updated the tests
- And consolidated a couple of individual ones, pulling the test data into the data provider, i.e. for consistency

File: tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php [hunk 4]
```diff
@@ -213,6 +213,72 @@ public function test_detect_cannot_create_folder() {
...
+
+		wp_privacy_generate_personal_data_export_file( self::$export_request_id );
+	}
+
+	public function data_export_data_grouped_invalid_type() {
+		return array(
+			array( 10 ),
+			array( '10' ),
+			array( true ),
+			array( new stdClass() ),
```
Comment (hellofromtonya) [inline tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php:236]: - Add `null` 
- Add `false`
- Change `'10'` to a `'foo'`
- Add serialized data string

Comment (hellofromtonya) [reply tests/phpunit/tests/privacy/wpPrivacyGeneratePersonalDataExportFile.php:238]: Resolved in commit https://github.com/WordPress/wordpress-develop/pull/758/commits/11361cad5a0e204c5e1ee7274ef0156d7e308d51

**Review by jrfnl — APPROVED** [review]
LGTM - discussed multiple times and does what we concluded should be done.
Tests also had significant improvements, so all together a great step forward.

**Comment by hellofromtonya** [comment]
Merged via changeset https://core.trac.wordpress.org/changeset/50613.

**Pull request closed without merging**